### PR TITLE
[operators] Add friend foreach_monomial()

### PR DESCRIPTION
**Pull request by krivenko**
Following suggestion of @francescopt.

https://github.com/TRIQS/triqs/issues/364#issuecomment-260948562

### CODE REVIEW

File: triqs/operators/many_body_operator.hpp [hunk 1]
```diff
@@ -285,28 +285,32 @@ namespace operators {
    return res;
   }
 
-  // real
-  friend many_body_operator_generic real(many_body_operator_generic const& op) {
+  // Transform operator by applying a given function to each monomial
+  template <typename Lambda>
+  friend many_body_operator_generic foreach_monomial(many_body_operator_generic const& op, Lambda L) {
```
Comment (Wentzell) [inline triqs/operators/many_body_operator.hpp]: Pass L as ``` Lambda const&```?

File: triqs/operators/many_body_operator.hpp [hunk 2]
```diff
@@ -285,28 +285,32 @@ namespace operators {
...
    many_body_operator_generic res;
-   using triqs::utility::real;
    using triqs::utility::is_zero;
    for (auto const& x : op) {
-    auto c = real(x.coef);
+    auto c = L(x.monomial, x.coef);
```
Comment (Wentzell) [inline triqs/operators/many_body_operator.hpp]: What do you have in mind that might change the coeff depending on the monomial as well?

File: triqs/operators/many_body_operator.hpp [hunk 3]
```diff
@@ -285,28 +285,32 @@ namespace operators {
...
     if(!is_zero(c)) res.monomials.insert({x.monomial, c});
    }
    return res;
   }
 
+  // real
+  friend many_body_operator_generic real(many_body_operator_generic const& op) {
+   return foreach_monomial(op, [](monomial_t const&, scalar_t c) {
+    using triqs::utility::real;
+    return real(c);
```
Comment (Wentzell) [inline triqs/operators/many_body_operator.hpp]: maybe just
```c++
return triqs::utility::real(c);
```

File: triqs/operators/many_body_operator.hpp (hunk 1 above)
Comment (krivenko) [inline triqs/operators/many_body_operator.hpp]: Good point. I think a universal reference would make even more sense here.

File: triqs/operators/many_body_operator.hpp (hunk 2 above)
Comment (krivenko) [inline triqs/operators/many_body_operator.hpp]: Any sort of filtration algorithm that returns 0 for each unwanted monomial.

File: triqs/operators/many_body_operator.hpp (hunk 3 above)
Comment (krivenko) [inline triqs/operators/many_body_operator.hpp]: Compilation of your variant will fail if `triqs::utility::real(scalar_t)` is not defined.
There is still a chance, however, that `real(scalar_t)` is found by argument-dependent lookup, which makes compilation of `real(c)` possible.

**Comment by krivenko** [comment]
O.k., so you want more generic names...

I am fine with `transform`, but `filter` is a special case of `transform` and thus redundant.
These two operations are essentially different for lists: `transform` preserves length of the list, while `filter` does not. For monomials, however, it's equivalent to get a zero coefficient or no coefficient at all.

**Comment by krivenko** [comment]
Just in case, I made the requested change...

**Comment by francescopt** [comment]
Another useful ```transform``` routine would consist in allowing a transformation of the indices. That is, you provide a lambda function ```indices_t->indices_t```, and iterate over the monomials and, for each monomial, over the indices. In principle one can implement such a function with the ```const``` access provided by the iterators, but it should be much more efficient if one introduces a friend function like the ```transform``` outlined above, because then one has direct write access to the monomials and to the ```normalize_and_insert``` routine.

**Pull request merged**